linera-core/src/client.rs

```diff
@@ -568,6 +629,96 @@ where
         Ok(())
     }
 
+    async fn synchronize_received_certificates_from_validator<A>(
+        chain_id: ChainId,
+        name: ValidatorName,
+        tracker: usize,
+        committees: BTreeMap<Epoch, Committee>,
+        max_epoch: Epoch,
+        mut client: A,
+    ) -> Result<(ValidatorName, usize, Vec<Certificate>), NodeError>
+    where
+        A: ValidatorNode + Send + Sync + 'static + Clone,
+    {
+        // Retrieve newly received certificates from this validator.
+        let query = ChainInfoQuery::new(chain_id).with_received_log_excluding_first_nth(tracker);
+        let response = client.handle_chain_info_query(query).await?;
+        // Responses are authenticated for accountability.
+        response.check(name)?;
+        let mut certificates = Vec::new();
+        let mut new_tracker = tracker;
+        for entry in response.info.requested_received_log {
+            let query = ChainInfoQuery::new(entry.chain_id)
+                .with_sent_certificates_in_range(BlockHeightRange::single(entry.height));
+
+            let mut response = client.handle_chain_info_query(query).await?;
+            let Some(certificate) = response.info
+                            .requested_sent_certificates.pop() else {
+                            break;
+                        };
```

Comment:
`rustfmt` doesn't want to touch `let … else` yet.

```suggestion
            let Some(certificate) = response.info
                .requested_sent_certificates.pop() else {
                break;
            };
```